File: react-native-sdk/ios/SecureSDKModule.swift
import Foundation
import React
import CryptoKit

/**
 * React Native iOS bridge — **thin wrapper only**.
 *
 * All crypto, keys, and envelopes are implemented in `SecuritySdk` (Swift sources under `ios-sdk/`),
 * compiled into this pod. Do not add business logic here; keep a single source of truth in the SDK.
 *
 * Objective-C (`SecureSDKBridge.m`) declares `RCT_EXTERN_MODULE` only — no duplicate implementations.
 */
@objc(SecureSDK)
final class SecureSDK: NSObject {
    private let maxBodyBytes = 256 * 1024
    private let maxKeyChars = 8 * 1024
    private var pinnedHost: String?
    private var pinnedPins: [String] = []

    @objc static func requiresMainQueueSetup() -> Bool {
        false
    }

    private func rejectSafe(_ reject: RCTPromiseRejectBlock, code: String, _ error: Error?) {
        // Do not pass raw exceptions / stacks to JS.
        let msg = error?.localizedDescription ?? "failed"
        reject(code, msg, nil)
    }

    @objc func initialize(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        do {
            try SecuritySdk.initSdk()
            resolve(nil)
        } catch {
            rejectSafe(reject, code: "E_INIT", error)
        }
    }

    @objc func configureServerPublicKey(_ base64: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        if base64.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty || base64.count > maxKeyChars {
            reject("E_CONFIG", "invalid server public key", nil)
            return
        }
        do {
            try SecuritySdk.configureServerPublicKeyFromBase64(base64)
            resolve(nil)
        } catch {
            rejectSafe(reject, code: "E_CONFIG", error)
        }
    }

    @objc func configurePinning(_ configJson: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        if configJson.count > maxBodyBytes {
            reject("E_PIN", "config too large", nil)
            return
        }
        guard let data = configJson.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let host = obj["host"] as? String,
              let pins = obj["pins"] as? [String],
              !host.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              !pins.isEmpty else {
            reject("E_PIN", "invalid pinning config", nil)
            return
        }
        self.pinnedHost = host
        self.pinnedPins = pins
        resolve(nil)
    }

    @objc func getPublicKey(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        do {
            resolve(try SecuritySdk.getPublicKey())
        } catch {
            rejectSafe(reject, code: "E_PUBLIC_KEY", error)
        }
    }

    @objc func rotateKeys(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        do {
            try SecuritySdk.rotateKeys()
            resolve(nil)
        } catch {
            rejectSafe(reject, code: "E_ROTATE", error)
        }
    }

    @objc func getDeviceRegistrationPayload(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        do {
            let p = try SecuritySdk.getDeviceRegistrationPayload()
            resolve(try Self.jsonString(from: p))
        } catch {
            rejectSafe(reject, code: "E_REG", error)
        }
    }

    @objc func secureRequest(_ path: String, bodyJson: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        if path.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            reject("E_REQUEST", "invalid path", nil)
            return
        }
        if let d = bodyJson.data(using: .utf8), d.count > maxBodyBytes {
            reject("E_REQUEST", "body too large", nil)
            return
        }
        do {
            guard let data = bodyJson.data(using: .utf8),
                  let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                reject("E_REQUEST", "Invalid JSON body", nil)
                return
            }
            let res = try SecuritySdk.secureRequest(path: path, body: obj)
            let payload: [String: Any] = [
                "statusCode": res.statusCode,
                "headers": res.headers,
                "body": res.body,
            ]
            resolve(try Self.jsonString(from: payload))
        } catch {
            rejectSafe(reject, code: "E_REQUEST", error)
        }
    }

    @objc func pinnedPost(_ url: String, headersJson: String, bodyJson: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        do {
            let raw = try pinnedHttpPost(url: url, headersJson: headersJson, bodyJson: bodyJson)
            resolve(raw)
        } catch {
            rejectSafe(reject, code: "E_HTTP", error)
        }
    }

    @objc func secureRequestPinned(_ url: String, bodyJson: String, stepUpToken: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        do {
            guard let data = bodyJson.data(using: .utf8),
                  let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                reject("E_REQUEST", "Invalid JSON body", nil)
                return
            }
            let envRes = try SecuritySdk.secureRequest(path: url, body: obj)
            // Extract local-only DEK and strip it before network forwarding.
            let dekB64 = (envRes.body["_clientDek"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
            var bodyFiltered = envRes.body
            bodyFiltered.removeValue(forKey: "_clientDek")
            let payload: [String: Any] = [
                "statusCode": envRes.statusCode,
                "headers": envRes.headers,
                "body": bodyFiltered,
            ]
            let rawBody = try Self.jsonString(from: payload)
            var headers: [String: String] = ["Content-Type": "application/json"]
            if !stepUpToken.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                headers["X-StepUp-Token"] = stepUpToken
            }
            let raw = try pinnedHttpPost(url: url, headersJson: try Self.jsonString(from: headers), bodyJson: rawBody)
            // Expect encrypted gateway response; decrypt locally and return plaintext SecureResponse shape.
            if let d = raw.data(using: .utf8),
               let respObj = try? JSONSerialization.jsonObject(with: d) as? [String: Any],
               let ok = respObj["ok"] as? Bool, ok == true,
               let enc = respObj["enc"] as? [String: Any],
               let iv = enc["aesIv"] as? String,
               let ct = enc["ciphertext"] as? String,
               let tag = enc["aesTag"] as? String,
               let dek = dekB64,
               let plain = try? decryptAesGcm(keyB64: dek, ivB64: iv, ctB64: ct, tagB64: tag),
               let plainData = plain.data(using: .utf8),
               let plainJson = try? JSONSerialization.jsonObject(with: plainData) as? [String: Any] {
                let out: [String: Any] = [
                    "statusCode": respObj["statusCode"] as? Int ?? 200,
                    "headers": respObj["headers"] as? [String: Any] ?? [:],
                    "body": plainJson,
                ]
                resolve(try Self.jsonString(from: out))
                return
            }
            resolve(raw)
        } catch {
            rejectSafe(reject, code: "E_REQUEST", error)
        }
    }

    @objc func signStepUp(_ message: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        if message.count > maxBodyBytes {
            reject("E_SIGN", "message too large", nil)
            return
        }
        do {
            let sig = try SecuritySdk.sign(data: Data(message.utf8))
            resolve(sig.base64EncodedString())
        } catch {
            rejectSafe(reject, code: "E_SIGN", error)
        }
    }

    private func pinnedHttpPost(url: String, headersJson: String, bodyJson: String) throws -> String {
        guard let host = pinnedHost, !pinnedPins.isEmpty else {
            throw NSError(domain: "SecureSDK", code: 0, userInfo: [NSLocalizedDescriptionKey: "pinning not configured"])
        }
        guard let u = URL(string: url), u.scheme?.lowercased() == "https", u.host?.lowercased() == host.lowercased() else {
            throw NSError(domain: "SecureSDK", code: 0, userInfo: [NSLocalizedDescriptionKey: "https required / host mismatch"])
        }
        guard let body = bodyJson.data(using: .utf8) else {
            throw NSError(domain: "SecureSDK", code: 0, userInfo: [NSLocalizedDescriptionKey: "utf8 encode failed"])
        }
        if body.count > maxBodyBytes {
            throw NSError(domain: "SecureSDK", code: 0, userInfo: [NSLocalizedDescriptionKey: "body too large"])
        }

        var req = URLRequest(url: u)
        req.httpMethod = "POST"
        req.httpBody = body

        if let hData = headersJson.data(using: .utf8),
           let hObj = try? JSONSerialization.jsonObject(with: hData) as? [String: Any] {
            for (k, v) in hObj {
                if let s = v as? String, !s.isEmpty {
                    req.setValue(s, forHTTPHeaderField: k)
                }
            }
        }

        let cfg = URLSessionConfiguration.ephemeral
        let session = URLSession(
            configuration: cfg,
            delegate: PinnedURLSessionDelegate(allowedSpkiSha256Base64: pinnedPins),
            delegateQueue: nil
        )

        let sem = DispatchSemaphore(value: 0)
        var outData: Data?
        var outResp: URLResponse?
        var outErr: Error?
        session.dataTask(with: req) { d, r, e in
            outData = d
            outResp = r
            outErr = e
            sem.signal()
        }.resume()
        _ = sem.wait(timeout: .now() + 20)

        if let e = outErr { throw e }
        guard let http = outResp as? HTTPURLResponse else {
            throw NSError(domain: "SecureSDK", code: 0, userInfo: [NSLocalizedDescriptionKey: "no response"])
        }
        guard (200...299).contains(http.statusCode) else {
            throw NSError(domain: "SecureSDK", code: http.statusCode, userInfo: [NSLocalizedDescriptionKey: "http_\(http.statusCode)"])
        }
        return String(data: outData ?? Data(), encoding: .utf8) ?? ""
    }

    @objc func getDeviceId(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        do {
            resolve(try SecuritySdk.getDeviceId())
        } catch {
            rejectSafe(reject, code: "E_DEVICE_ID", error)
        }
    }

    @objc func getSecurityStatus(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        let s = SecuritySdk.getSecurityStatus()
        let dict: [String: Any] = [
            "riskScore": s.riskScore,
            "findings": s.findings,
        ]
        do {
            resolve(try Self.jsonString(from: dict))
        } catch {
            rejectSafe(reject, code: "E_STATUS", error)
        }
    }

    private static func jsonString(from object: [String: Any]) throws -> String {
        let d = try JSONSerialization.data(withJSONObject: object, options: [])
        guard let s = String(data: d, encoding: .utf8) else {
            throw NSError(domain: "SecureSDK", code: 0, userInfo: [NSLocalizedDescriptionKey: "UTF-8 encode failed"])
        }
        return s
    }

    private func decryptAesGcm(keyB64: String, ivB64: String, ctB64: String, tagB64: String) throws -> String {
        guard let key = Data(base64Encoded: keyB64),
              let iv = Data(base64Encoded: ivB64),
              let ct = Data(base64Encoded: ctB64),
              let tag = Data(base64Encoded: tagB64) else {
            throw NSError(domain: "SecureSDK", code: 0, userInfo: [NSLocalizedDescriptionKey: "invalid enc fields"])
        }
        let sealed = try AES.GCM.SealedBox(nonce: AES.GCM.Nonce(data: iv), ciphertext: ct, tag: tag)
        let plain = try AES.GCM.open(sealed, using: SymmetricKey(data: key))
        return String(data: plain, encoding: .utf8) ?? ""
    }
}

File: react-native-sdk/android/src/main/kotlin/com/bankingsdk/security/rn/SecureSDKModule.kt
package com.bankingsdk.security.rn

import com.bankingsdk.security.SecuritySdk
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import okhttp3.CertificatePinner
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import android.util.Base64
import org.json.JSONObject
import java.net.URL
import java.util.concurrent.TimeUnit
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

class SecureSDKModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String = "SecureSDK"

    private val maxBodyChars = 256 * 1024
    private val maxKeyChars = 8 * 1024
    @Volatile private var pinnedHost: String? = null
    @Volatile private var pinnedPins: List<String> = emptyList() // base64 pins (no sha256/ prefix)

    private fun rejectSafe(promise: Promise, code: String, t: Throwable?) {
        // Do not pass raw exceptions / stacks to JS.
        val msg = t?.message ?: "failed"
        promise.reject(code, msg, null)
    }

    @ReactMethod
    fun initialize(promise: Promise) {
        val result = SecuritySdk.init(reactApplicationContext.applicationContext)
        result.fold(
            onSuccess = { promise.resolve(null) },
            onFailure = { rejectSafe(promise, "E_INIT", it) },
        )
    }

    @ReactMethod
    fun getPublicKey(promise: Promise) {
        val result = SecuritySdk.getPublicKey()
        result.fold(
            onSuccess = { promise.resolve(it) },
            onFailure = { rejectSafe(promise, "E_PUBLIC_KEY", it) },
        )
    }

    @ReactMethod
    fun rotateKeys(promise: Promise) {
        val result = SecuritySdk.rotateKeys()
        result.fold(
            onSuccess = { promise.resolve(null) },
            onFailure = { rejectSafe(promise, "E_ROTATE", it) },
        )
    }

    @ReactMethod
    fun getDeviceRegistrationPayload(promise: Promise) {
        val result = SecuritySdk.getDeviceRegistrationPayload()
        result.fold(
            onSuccess = { map ->
                val o = JSONObject()
                map.forEach { (k, v) -> o.put(k, v) }
                promise.resolve(o.toString())
            },
            onFailure = { rejectSafe(promise, "E_REG", it) },
        )
    }

    @ReactMethod
    fun configureServerPublicKey(base64: String, promise: Promise) {
        if (base64.isBlank() || base64.length > maxKeyChars) {
            promise.reject("E_CONFIG", "invalid server public key", null)
            return
        }
        val result = SecuritySdk.configureServerPublicKeyFromBase64(base64)
        result.fold(
            onSuccess = { promise.resolve(null) },
            onFailure = { rejectSafe(promise, "E_CONFIG", it) },
        )
    }

    @ReactMethod
    fun configurePinning(configJson: String, promise: Promise) {
        if (configJson.length > maxBodyChars) {
            promise.reject("E_PIN", "config too large", null)
            return
        }
        val obj = try { JSONObject(configJson) } catch (_: Throwable) {
            promise.reject("E_PIN", "invalid JSON", null)
            return
        }
        val host = obj.optString("host", "").trim()
        val pinsArr = obj.optJSONArray("pins")
        if (host.isBlank() || pinsArr == null || pinsArr.length() == 0) {
            promise.reject("E_PIN", "invalid pinning config", null)
            return
        }
        val pins = mutableListOf<String>()
        for (i in 0 until pinsArr.length()) {
            val p = pinsArr.optString(i, "").trim()
            if (p.isBlank() || p.length > maxKeyChars) {
                promise.reject("E_PIN", "invalid pin", null)
                return
            }
            pins.add(p)
        }
        pinnedHost = host
        pinnedPins = pins
        promise.resolve(null)
    }

    @ReactMethod
    fun pinnedPost(url: String, headersJson: String, bodyJson: String, promise: Promise) {
        if (url.isBlank()) {
            promise.reject("E_HTTP", "invalid url", null)
            return
        }
        if (headersJson.length > maxBodyChars || bodyJson.length > maxBodyChars) {
            promise.reject("E_HTTP", "payload too large", null)
            return
        }
        try {
            val res = doPinnedPost(url, headersJson, bodyJson)
            promise.resolve(res)
        } catch (t: Throwable) {
            rejectSafe(promise, "E_HTTP", t)
        }
    }

    @ReactMethod
    fun secureRequestPinned(url: String, bodyJson: String, stepUpToken: String?, promise: Promise) {
        if (url.isBlank()) {
            promise.reject("E_REQUEST", "invalid url", null)
            return
        }
        if (bodyJson.length > maxBodyChars) {
            promise.reject("E_REQUEST", "body too large", null)
            return
        }
        val bodyObj = try { JSONObject(bodyJson) } catch (_: Throwable) {
            promise.reject("E_REQUEST", "invalid JSON body", null)
            return
        }
        val bodyMap = jsonObjectToMap(bodyObj)
        val env = SecuritySdk.secureRequest(url, bodyMap)
        env.fold(
            onSuccess = { secureRes ->
                val out = JSONObject()
                out.put("statusCode", secureRes.statusCode)
                val headersOut = JSONObject()
                secureRes.headers.forEach { (k, v) -> headersOut.put(k, v) }
                out.put("headers", headersOut)
                // Extract local-only DEK and strip it before network forwarding.
                val dekB64 = (secureRes.body["_clientDek"] as? String)?.trim()
                val bodyFiltered = secureRes.body.toMutableMap().also { it.remove("_clientDek") }
                out.put("body", mapToJsonObject(bodyFiltered))
                try {
                    val headers = JSONObject()
                    headers.put("Content-Type", "application/json")
                    if (!stepUpToken.isNullOrBlank()) headers.put("X-StepUp-Token", stepUpToken)
                    val raw = doPinnedPost(url, headers.toString(), out.toString())
                    // Expect encrypted gateway response; decrypt locally and return plaintext SecureResponse shape.
                    val rawObj = JSONObject(raw)
                    val ok = rawObj.optBoolean("ok", false)
                    if (!ok) {
                        promise.resolve(raw) // pass through error JSON
                        return@fold
                    }
                    val enc = rawObj.optJSONObject("enc")
                    if (enc == null || dekB64.isNullOrBlank()) {
                        promise.resolve(raw)
                        return@fold
                    }
                    val plain = decryptAesGcm(
                        dekB64,
                        enc.optString("aesIv", ""),
                        enc.optString("ciphertext", ""),
                        enc.optString("aesTag", ""),
                    )
                    val resp = JSONObject()
                    resp.put("statusCode", rawObj.optInt("statusCode", 200))
                    resp.put("headers", rawObj.optJSONObject("headers") ?: JSONObject())
                    resp.put("body", JSONObject(plain))
                    promise.resolve(resp.toString())
                } catch (t: Throwable) {
                    rejectSafe(promise, "E_REQUEST", t)
                }
            },
            onFailure = { rejectSafe(promise, "E_REQUEST", it) },
        )
    }

    @ReactMethod
    fun signStepUp(message: String, promise: Promise) {
        if (message.length > maxBodyChars) {
            promise.reject("E_SIGN", "message too large", null)
            return
        }
        val res = SecuritySdk.sign(message.toByteArray(Charsets.UTF_8))
        res.fold(
            onSuccess = { sig ->
                promise.resolve(Base64.encodeToString(sig, Base64.NO_WRAP))
            },
            onFailure = { rejectSafe(promise, "E_SIGN", it) },
        )
    }

    @ReactMethod
    fun secureRequest(path: String, bodyJson: String, promise: Promise) {
        if (path.isBlank()) {
            promise.reject("E_REQUEST", "invalid path", null)
            return
        }
        if (bodyJson.length > maxBodyChars) {
            promise.reject("E_REQUEST", "body too large", null)
            return
        }
        val bodyObj = try {
            JSONObject(bodyJson)
        } catch (_: Throwable) {
            promise.reject("E_REQUEST", "invalid JSON body", null)
            return
        }
        val bodyMap = jsonObjectToMap(bodyObj)
        val result = SecuritySdk.secureRequest(path, bodyMap)
        result.fold(
            onSuccess = { res ->
                val out = JSONObject()
                out.put("statusCode", res.statusCode)
                val headersJson = JSONObject()
                res.headers.forEach { (k, v) -> headersJson.put(k, v) }
                out.put("headers", headersJson)
                out.put("body", mapToJsonObject(res.body))
                promise.resolve(out.toString())
            },
            onFailure = { rejectSafe(promise, "E_REQUEST", it) },
        )
    }

    @ReactMethod
    fun getDeviceId(promise: Promise) {
        val result = SecuritySdk.getDeviceId()
        result.fold(
            onSuccess = { promise.resolve(it) },
            onFailure = { rejectSafe(promise, "E_DEVICE_ID", it) },
        )
    }

    @ReactMethod
    fun getSecurityStatus(promise: Promise) {
        val s = SecuritySdk.getSecurityStatus()
        val out = JSONObject()
        out.put("riskScore", s.riskScore)
        out.put("findings", org.json.JSONArray(s.findings))
        promise.resolve(out.toString())
    }

    private fun jsonObjectToMap(obj: JSONObject): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>()
        val keys = obj.keys()
        while (keys.hasNext()) {
            val k = keys.next()
            map[k] = obj.get(k).takeUnless { it === JSONObject.NULL }
        }
        return map
    }

    private fun mapToJsonObject(body: Map<String, Any?>): JSONObject {
        val o = JSONObject()
        body.forEach { (k, v) ->
            when (v) {
                null -> o.put(k, JSONObject.NULL)
                is Map<*, *> -> o.put(k, mapToJsonObject(v as Map<String, Any?>))
                is Long -> o.put(k, v)
                is Int -> o.put(k, v)
                is Double -> o.put(k, v)
                is Float -> o.put(k, v.toDouble())
                is Boolean -> o.put(k, v)
                is String -> o.put(k, v)
                else -> o.put(k, v.toString())
            }
        }
        return o
    }

    private fun buildPinnedClient(host: String, pins: List<String>): OkHttpClient {
        val pinnerBuilder = CertificatePinner.Builder()
        pins.forEach { pin ->
            // OkHttp expects `sha256/<base64>`
            pinnerBuilder.add(host, "sha256/$pin")
        }
        return OkHttpClient.Builder()
            .certificatePinner(pinnerBuilder.build())
            .connectTimeout(10, TimeUnit.SECONDS)
            .readTimeout(15, TimeUnit.SECONDS)
            .build()
    }

    private fun doPinnedPost(url: String, headersJson: String, bodyJson: String): String {
        val host = pinnedHost ?: throw IllegalStateException("pinning not configured")
        val pins = pinnedPins
        if (pins.isEmpty()) throw IllegalStateException("pinning not configured")

        val parsed = URL(url)
        if (!parsed.protocol.equals("https", ignoreCase = true)) throw IllegalArgumentException("https required")
        if (!parsed.host.equals(host, ignoreCase = true)) throw IllegalArgumentException("host mismatch")

        val headersObj = try { JSONObject(headersJson) } catch (_: Throwable) { JSONObject() }
        val reqBuilder = Request.Builder().url(url)
        val keys = headersObj.keys()
        while (keys.hasNext()) {
            val k = keys.next()
            val v = headersObj.optString(k, "")
            if (k.isNotBlank() && v.isNotBlank()) reqBuilder.addHeader(k, v)
        }

        val media = "application/json; charset=utf-8".toMediaType()
        val req = reqBuilder.post(bodyJson.toRequestBody(media)).build()
        val client = buildPinnedClient(host, pins)
        client.newCall(req).execute().use { resp ->
            val body = resp.body?.string() ?: ""
            if (!resp.isSuccessful) throw IllegalStateException("http_${resp.code}")
            return body
        }
    }

    private fun decryptAesGcm(keyB64: String, ivB64: String, ctB64: String, tagB64: String): String {
        val key = Base64.decode(keyB64, Base64.NO_WRAP)
        val iv = Base64.decode(ivB64, Base64.NO_WRAP)
        val ct = Base64.decode(ctB64, Base64.NO_WRAP)
        val tag = Base64.decode(tagB64, Base64.NO_WRAP)
        val combined = ByteArray(ct.size + tag.size)
        System.arraycopy(ct, 0, combined, 0, ct.size)
        System.arraycopy(tag, 0, combined, ct.size, tag.size)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, iv))
        val plain = cipher.doFinal(combined)
        return String(plain, Charsets.UTF_8)
    }
}
